From 7237238de8314e90010728441efa29e6f94d6d2a Mon Sep 17 00:00:00 2001
From: Noam Lessner <nlessner@habana.ai>
Date: Sun, 16 Feb 2020 08:55:53 +0200
Subject: [PATCH] [SW-8229] goya2: add progress report of rom

Report to PSOC_SECURITY the general progress through
the stages of ROM.

Change-Id: I42b72fe481bd7251c78343e857f16e93c471d0cd
Signed-off-by: Noam Lessner <nlessner@habana.ai>
---

diff --git a/app/arcboot_rom.c b/app/arcboot_rom.c
index a4a740a..727fdc9 100644
--- a/app/arcboot_rom.c
+++ b/app/arcboot_rom.c
@@ -23,9 +23,40 @@
 #define ARCBOOT_BRINGUP_ERROR		-1
 #define ARCBOOT_BRINGUP_OK		1
 
+void arcboot_rom_status_update(enum arcboot_rom_progress progress)
+{
+	uint32_t reg;
+
+	reg = regmap_readl(mmPSOC_SECURITY_SW_STATUS_0);
+	reg |= BIT(progress);
+	regmap_writel(reg, mmPSOC_SECURITY_SW_STATUS_0);
+}
+
+static void arcboot_rom_status_update_clear(uint8_t mask,
+				     uint8_t position)
+{
+	uint32_t reg;
+
+	reg = regmap_readl(mmPSOC_SECURITY_SW_STATUS_0);
+	reg &= (~mask) << position;
+	regmap_writel(reg, mmPSOC_SECURITY_SW_STATUS_0);
+}
+
+void arcboot_rom_status_update_location(uint8_t progress,
+					enum security_location loc)
+{
+	uint32_t reg;
+
+	reg = regmap_readl(mmPSOC_SECURITY_SW_STATUS_0);
+	reg |= loc << progress;
+	regmap_writel(reg, mmPSOC_SECURITY_SW_STATUS_0);
+}
+
 static void arcboot_rom_init(void)
 {
 	uint32_t reg;
+
+	arcboot_rom_status_update(PROGRESS_INIT);
 
 	/* make sure spi controller doesn't swap bytes */
 	spi_flash_set_byte_swap(DISABLE_BYTE_SWAP);
@@ -141,6 +172,8 @@
 	uint64_t pcie_fw_offset;
 	uint64_t pcie_fw_dest;
 
+	arcboot_rom_status_update_clear(CLEAR_DMA_SECURITY_PROGRESS,
+					PROGRESS_BRINGUP_PCI_DMA);
 	flash_image_offset = arcboot_flash_image_offset(image_ptr);
 	flash_ppboot_pcie_offset = (uint32_t)(FLASH_64KB_PADD +
 					      flash_image_offset);
@@ -152,6 +185,8 @@
 		/* TODO: update PSOC_SECURITY about CRC error in image */
 		return ARCBOOT_BRINGUP_ERROR;
 
+	arcboot_rom_status_update(PROGRESS_BRINGUP_PCI_DMA);
+
 	/* remove headers and footers from image, this is the image to verify */
 	ppboot_pcie_size -= SECUREBOOT_IMG_HEAD_FOOTER;
 	/* offset of ppboot+pcie_fw image to verify, start from fw_version */
@@ -160,6 +195,8 @@
 	    SECUREBOOT_VERIFY_FAILED)
 		/* TODO: update PSOC_SECURITY about verify fail in image */
 		return ARCBOOT_BRINGUP_ERROR;
+
+	arcboot_rom_status_update(PROGRESS_BRINGUP_PCI_SECURITY);
 
 	ppboot_size = fw_sram_readl(PPBOOT_IMAGE_SIZE_OFFSET);
 
@@ -173,6 +210,8 @@
 	pcie_fw_offset = HL_SRAM_BASE + HL_FWSRAM_OFFSET + ppboot_size;
 	/* address of pcie_fw_sram in sram space, as destination of mem_cpy */
 	pcie_fw_dest = HL_SRAM_BASE + HL_PCIE_FWSRAM_OFFSET;
+
+	arcboot_rom_status_update(PROGRESS_BRINGUP_PCI_MEM_CPY);
 
 	/* copy to pcie_fw_sram from fw_sram the pcie_fw */
 	mem_cpy(pcie_fw_size, pcie_fw_offset, pcie_fw_dest);
@@ -188,9 +227,11 @@
 {
 	void (*ppboot)();
 
-	if (arcboot_rom_fw_is_not_validated())
+	if (arcboot_rom_fw_is_not_validated()) {
+		arcboot_rom_status_update(PROGRESS_BRINGUP_PCI_PRI);
 		if (arcboot_rom_bringup_pcie_images(
 		    ARCBOOT_PRIMARY) != ARCBOOT_BRINGUP_OK) {
+			arcboot_rom_status_update(PROGRESS_BRINGUP_PCI_SEC);
 			if (arcboot_rom_bringup_pcie_images(
 			    ARCBOOT_SECONDARY) != ARCBOOT_BRINGUP_OK) {
 				arcboot_rom_validation_fail();
@@ -200,11 +241,15 @@
 					;
 			}
 		}
+	}
+	arcboot_rom_status_update(PROGRESS_PPBOOT_START);
+
 	/* cast ppboot location as function and execute it */
 	ppboot = (void *)(HL_FWSRAM_BASE + PPBOOT_IMAGE_CODE_OFFSET);
 	ppboot();
 
 	arcboot_rom_clear_recovery_flag();
+	arcboot_rom_status_update(PROGRESS_PPBOOT_FINISH);
 }
 
 static int arcboot_rom_bringup_ram_images(uint8_t image_ptr)
@@ -212,6 +257,8 @@
 	uint32_t flash_arcboot_ram_offset;
 	uint16_t arcboot_ram_size;
 
+	arcboot_rom_status_update_clear(CLEAR_DMA_SECURITY_PROGRESS,
+					PROGRESS_BRINGUP_RAM_DMA);
 	flash_arcboot_ram_offset = arcboot_flash_image_offset(image_ptr);
 	arcboot_ram_size = (uint16_t)spi_flash_readl(flash_arcboot_ram_offset);
 
@@ -219,19 +266,24 @@
 		       HL_SRAM_BASE + HL_ICCM_OFFSET, SPI_DMA_DST_HBW);
 	if (dma_wait_finish() == DMA_CRC_ERROR)
 		return ARCBOOT_BRINGUP_ERROR;
+	arcboot_rom_status_update(PROGRESS_BRINGUP_RAM_DMA);
 
 	arcboot_ram_size -= SECUREBOOT_IMG_HEAD_FOOTER;
 	if (secureboot_verify_image(arcboot_ram_size, IMG_LEN_SIZE) ==
 	    SECUREBOOT_VERIFY_FAILED)
 		return ARCBOOT_BRINGUP_ERROR;
+	arcboot_rom_status_update(PROGRESS_BRINGUP_RAM_SECURITY);
 
 	return ARCBOOT_BRINGUP_OK;
 }
 
 static void arcboot_rom_bringup_arcboot_ram(void)
 {
+
+	arcboot_rom_status_update(PROGRESS_BRINGUP_RAM_PRI);
 	if (arcboot_rom_bringup_ram_images(ARCBOOT_PRIMARY) !=
 	    ARCBOOT_BRINGUP_OK) {
+		arcboot_rom_status_update(PROGRESS_BRINGUP_RAM_SEC);
 		if (arcboot_rom_bringup_ram_images(ARCBOOT_SECONDARY) !=
 		    ARCBOOT_BRINGUP_OK) {
 			/* if debug I/F is enabled then fall back on I2C_S */
@@ -241,6 +293,8 @@
 		}
 	}
 
+	arcboot_rom_status_update(PROGRESS_ARCBOOT_RAM_START);
+
 	/* jump to iccm start location */
 	goto *(void *)(0);
 }
diff --git a/common/pcie.c b/common/pcie.c
index 82a0e73..3be10ef 100644
--- a/common/pcie.c
+++ b/common/pcie.c
@@ -78,6 +78,8 @@
 	uint8_t spi_en = arcboot_rom_global.config_fuse &
 			EFUSE_SYSCFG_SPI_CONTROLLER;
 
+	arcboot_rom_status_update(PROGRESS_PCIE_IN_RECOVERY);
+
 	security_debug_if_secure();
 	security_debug_if_shared();
 
diff --git a/common/security.c b/common/security.c
index de43106..8eef54f 100644
--- a/common/security.c
+++ b/common/security.c
@@ -515,6 +515,10 @@
 			   EFUSE_SYSCFG_REVOKATION_LOCATION) &
 			   EFUSE_SYSCFG_REVOKATION_LOCATION_MASK);
 
+	arcboot_rom_status_update(PROGRESS_REVOKE_LOC);
+	arcboot_rom_status_update_location(PROGRESS_REVOKE_LOC_0,
+					   revoke_location);
+
 	switch (revoke_location) {
 	case SPI_FLASH:
 		spi_flash_get_revoke(&key_revoke);
@@ -547,6 +551,10 @@
 	fw_version_location = ((arcboot_rom_global.config_fuse >>
 			       EFUSE_SYSCFG_FW_VERSION_LOCATION) &
 			       EFUSE_SYSCFG_FW_VERSION_LOCATION_MASK);
+
+	arcboot_rom_status_update(PROGRESS_VER_LOC);
+	arcboot_rom_status_update_location(PROGRESS_VER_LOC_0,
+					   fw_version_location);
 
 	switch (fw_version_location) {
 	case SPI_FLASH:
@@ -583,6 +591,9 @@
 			EFUSE_SYSCFG_KEY_LOCATION) &
 			EFUSE_SYSCFG_KEY_LOCATION_MASK);
 
+	arcboot_rom_status_update(PROGRESS_KEY_LOC);
+	arcboot_rom_status_update_location(PROGRESS_KEY_LOC_0, key_location);
+
 	switch (key_location) {
 	case SPI_FLASH:
 		spi_flash_get_hash(key_hash);
diff --git a/common/security.h b/common/security.h
index 0e2295e..a7e3e91 100644
--- a/common/security.h
+++ b/common/security.h
@@ -81,7 +81,6 @@
 /* 16 bytes */
 #define SECURITY_EFUSE_KEYS_REVOKE_OFFSET	176
 
-
 void security_disable_debug_if(void);
 void security_debug_if_secure(void);
 void security_debug_if_shared(void);
diff --git a/include/arcboot.h b/include/arcboot.h
index 28379ac..63a2b5a 100644
--- a/include/arcboot.h
+++ b/include/arcboot.h
@@ -58,10 +58,34 @@
 	uint32_t config_fuse;
 	uint8_t recovery_flag;
 	uint8_t validation_pass;
+	uint8_t curr_img;
 };
 
-extern struct arcboot_global_s arcboot_rom_global;
-void hl_arcboot_rom_prstn_reset_config(void);
+enum arcboot_rom_progress {
+	PROGRESS_INIT,
+	PROGRESS_REVOKE_LOC,
+	PROGRESS_REVOKE_LOC_0,
+	PROGRESS_REVOKE_LOC_1,
+	PROGRESS_KEY_LOC,
+	PROGRESS_KEY_LOC_0,
+	PROGRESS_KEY_LOC_1,
+	PROGRESS_VER_LOC,
+	PROGRESS_VER_LOC_0,
+	PROGRESS_VER_LOC_1,		/* 10 */
+	PROGRESS_BRINGUP_PCI_PRI,
+	PROGRESS_BRINGUP_PCI_SEC,
+	PROGRESS_BRINGUP_PCI_DMA,
+	PROGRESS_BRINGUP_PCI_SECURITY,
+	PROGRESS_BRINGUP_PCI_MEM_CPY,
+	PROGRESS_PCIE_IN_RECOVERY,
+	PROGRESS_PPBOOT_START,
+	PROGRESS_PPBOOT_FINISH,
+	PROGRESS_BRINGUP_RAM_PRI,
+	PROGRESS_BRINGUP_RAM_SEC,	/* 20 */
+	PROGRESS_BRINGUP_RAM_DMA,
+	PROGRESS_BRINGUP_RAM_SECURITY,
+	PROGRESS_ARCBOOT_RAM_START
+};
 
 enum security_location {
 	ASIC_EFUSE,
@@ -69,6 +93,14 @@
 	SPI_FLASH
 };
 
+extern struct arcboot_global_s arcboot_rom_global;
+void hl_arcboot_rom_prstn_reset_config(void);
+void arcboot_rom_status_update(enum arcboot_rom_progress progress);
+void arcboot_rom_status_update_location(uint8_t progress,
+					enum security_location loc);
+
+#define CLEAR_DMA_SECURITY_PROGRESS 0x3
+
 static inline void base_writel(uint32_t val, uint32_t offset)
 {
 	*((volatile uint32_t *)(offset)) = val;
